[NVFUSER] refactor nvfuser build

jjsjann123 opened the pull request:
This PR is the first step towards refactors the build for nvfuser in order to have the coegen being a standalone library.

Contents inside this PR:
1. nvfuser code base has been moved to `./nvfuser`, from `./torch/csrc/jit/codegen/cuda/`, except for registration code for integration (interface.h/interface.cpp)
2. splits the build system so nvfuser is generating its own `.so` files. Currently there are:
    - `libnvfuser_codegen.so`, which contains the integration, codegen and runtime system of nvfuser
    - `nvfuser.so`, which is nvfuser's python API via pybind. Python frontend is now exposed via `nvfuser._C.XXX` instead of `torch._C._nvfuser`
3. nvfuser cpp tests is currently being compiled into `nvfuser_tests`
4. cmake is refactored so that:
    - nvfuser now has its own `CMakeLists.txt`, which is under `torch/csrc/jit/codegen/cuda/`.
    - nvfuser backend code is not compiled inside `libtorch_cuda_xxx` any more
    - nvfuser is added as a subdirectory under `./CMakeLists.txt` at the very end after torch is built.
    - since nvfuser has dependency on torch, the registration of nvfuser at runtime is done via dlopen (`at::DynamicLibrary`). This avoids circular dependency in cmake, which will be a nightmare to handle. For details, look at `torch/csrc/jit/codegen/cuda/interface.cpp::LoadingNvfuserLibrary`

Future work that's scoped in following PR:
- Currently since nvfuser codegen has dependency on torch, we need to refactor that out so we can move nvfuser into a submodule and not rely on dlopen to load the library. @malfet 
- Since we moved nvfuser into a cmake build, we effectively disabled bazel build for nvfuser. This could impact internal workload at Meta, so we need to put support back. cc'ing @vors


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang @kevinstephano @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

jjsjann123 committed:
stripping nvfuser codegen source

jjsjann123 committed:
added CMake files for nvfuser

jjsjann123 committed:
adding nvfuser lib

jjsjann123 committed:
adding tests

jjsjann123 committed:
adding pybind11

jjsjann123 committed:
updating pybind build

jjsjann123 committed:
fixing torch._C._nvfuser -> torch._C_nvfuser

jjsjann123 committed:
opt out of nvfuser build on non-cuda

jjsjann123 committed:
fixing enable flag

jjsjann123 committed:
fixing cmake find packages

jjsjann123 committed:
fixing build from scratch

jjsjann123 committed:
fixing signed/unsigned comparator

jjsjann123 committed:
hacky patch?!

jjsjann123 committed:
fixing signed/unsigned comparator

jjsjann123 committed:
cmake attempt again

jjsjann123 committed:
patching cmake for python::python

jjsjann123 committed:
fixing unsigned/signed comparator

jjsjann123 committed:
updating cmake dependencies

jjsjann123 committed:
attempt at patching build

jjsjann123 committed:
updating cache update on BUILD_NVFUSER

jjsjann123 committed:
fixing signed/unsigned comparator

jjsjann123 committed:
fixing more signed/unsigned for tests

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jjsjann123 committed:
disable unused_variable warning

jjsjann123 committed:
fixing MSVC c2491

jjsjann123 committed:
removing TORCH_CUDA_CU_API from definition

jjsjann123 committed:
fixing lazy initialization of cuda context

jjsjann123 committed:
fixing dllimport/dllexport for nvfuser

jjsjann123 committed:
Merge remote-tracking branch 'upstream/viable/strict' into HEAD

jjsjann123 committed:
fixing msvc build of missing nvtx include; adding bw compat allows list for removed nvfuser op registration

jjsjann123 commented:
tagging @vors on this PR per our conversation: https://pytorch.slack.com/archives/CUBDLTQ2G/p1669147594971219?thread_ts=1669145719.641899&cid=CUBDLTQ2G

Note that it's not ready to be reviewed yet. Also I totally didn't plumb things through bazel yet (plan to do that in follow up PRs). :laughing:

jjsjann123 commented:
macos failure seems to be flaky since it passed in earlier commits: https://github.com/pytorch/pytorch/actions/runs/3587956387/jobs/6038842972

jjsjann123 committed:
fixing NVFUSER_PYTHON libraries linking again

jjsjann123 committed:
fixing TORCH_API -> TORCH_CUDA_CU_API for libnvfuser_codegen

jjsjann123 committed:
fixing MSVC export on torch/csrc/jit/passes/pass_manager.h:PassManager

jjsjann123 committed:
fixing MSVC unresolved external symbol

jjsjann123 commented:
filed the issue on tests failure #90016

jjsjann123 commented:
looks like the libtorch_python.so unresolved symbol thing mysteriously went away :confused:

:crossed_fingers: I'll hope it remains that way. I'm gonna try get rocm build with nvfuser now.

jjsjann123 committed:
enable rocm build

jjsjann123 committed:
lintrunner

malfet commented on CMakeLists.txt:
```diff
@@ -185,6 +185,7 @@ option(USE_TSAN "Use Thread Sanitizer" OFF)
 option(USE_CUDA "Use CUDA" ON)
 cmake_dependent_option(
      BUILD_LAZY_CUDA_LINALG "Build cuda linalg ops as separate library" ON "USE_CUDA AND LINUX AND BUILD_PYTHON" OFF)
+option(BUILD_NVFUSER "Build NVFUSER" ON)
```
Can it be `cmake_dependent_option`
```suggestion
cmake_dependent_option(BUILD_NVFUSER "Build NVFUSER"  ON "USE_UCDA OR USE_ROCM" OFF)
```

jjsjann123 committed:
switching BUILD_NVFUSER to a dependent option on USE_CUDA OR USE_ROCM

jjsjann123 committed:
removing commented out build lines

jjsjann123 committed:
Merge remote-tracking branch 'upstream/viable/strict' into HEAD

jjsjann123 committed:
desparate attempt on python dependency

jjsjann123 committed:
fixing nvfuser python binding build

jjsjann123 committed:
Merge commit 'b95ea4f1494dacf86b5789f5466025b1fac1f2bb' into HEAD

jjsjann123 commented:
Phew~~ cleaned up. sorry about the spam.

jjsjann123 committed:
lintrunner

jjsjann123 committed:
lintrunner #2

jjsjann123 commented:
inductor failure looks irrelevant... rocm failure looks flaky.

msvc failure looks real. Guess my nvfuser extension thing doesn't work for win :angry:  Wondering what was done differently from functorch.....

jjsjann123 committed:
packaging nvfuser packages for win build

jjsjann123 committed:
moving files

jjsjann123 committed:
updating build & directory

jjsjann123 committed:
updating include

jjsjann123 committed:
moving cleaned up. things working locally now

jjsjann123 committed:
Merge commit 'f2587537992e873f29a7ff68870c37eef9a3fa53' into HEAD

jjsjann123 committed:
hipify enabled for nvfuser folder

jjsjann123 committed:
fixing win cpu build for packaging nvfuser

jjsjann123 commented on .jenkins/pytorch/common_utils.sh:
```diff
@@ -160,7 +160,7 @@ function checkout_install_torchdeploy() {
   local commit
   setup_torchdeploy_deps
   pushd ..
-  git clone --recurse-submodules https://github.com/pytorch/multipy.git
```
Oops, sorry about this. I thought I reverted this commit. Will do

atalman commented on nvfuser/csrc/compute_at_map.cpp:
```diff
@@ -431,7 +431,7 @@ void IterDomainGraph::build(Fusion* fusion) {
             // has an equivalent compute at leaf domain. For that case, we map
             // the equivalent compute at leaf domain.
-            for (int i = 0; i < p_tv->getComputeAtPosition(); i++) {
+            for (int i = 0; i < int(p_tv->getComputeAtPosition()); i++) {
```
nit: Why do we need to create new int here is the return from this method, unsigned int ? Maybe its better to use i as unsigned ?

jjsjann123 committed:
review comments; revert unwanted changes

davidberard98 commented on torch/csrc/jit/codegen/cuda/interface.cpp:
```diff
@@ -26,6 +26,32 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 
+#ifdef USE_CUDA
+class LoadingNvfuserLibrary {
```
I think this doesn't work internally:
- I don't think USE_CUDA is defined in the cuda builds? (cc @malfet, is this true? I remember running into this in the past with cudnn jit issues, I don't remember the resolution)
- We probably would need to add new buck targets for libnvfuser_codegen.so, and idk if "looking for a .so file at runtime" will work with buck?

fwiw, it appears that currently this fails silently internally (`setEnabled(true)` doesn't error out, but nvfuser doesn't run)

jjsjann123 replied:
Silent error is bad. I'll add an API to query whether nvfuser is loaded and inside setEnabled we'll check for it so at least it's not a silent error.

>    I don't think USE_CUDA is defined in the cuda builds? (cc @malfet, is this true? I remember running into this in the past with cudnn jit issues, I don't remember the resolution)

this is a surprise. Should I be looking for a different flag? We have USE_CUDA everywhere in the code tho, even under torch/... Wondering if there's a preprocessing or something that translates it shrug

>    We probably would need to add new buck targets for libnvfuser_codegen.so, and idk if "looking for a .so file at runtime" will work with buck?

So here's really two problems here:

1. buck targets for libnvfuser_codegen.so
    I understand it as buck relies on bazel file to build? I don't know if there's any extra hook to put it together tho.
    I do have a TODO to enable bazel in a following PR. We can drag that into this PR if we absolutely need bazel to work as well.

2. if "looking for a .so file at runtime" will work with buck?
    You guys are definitely the expert with your internal workflow. I'll follow your advise on how to proceed here.

    I was hoping it would work since we used it for nvrtc as well: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/detail/CUDAHooks.cpp#L174-L187

    But there is the USE_DIRECT_NVRTC macro that you guys might be using for internal workload.

    If dlopen is out of the window, is there an alternative for dynamically loading libraries for internal workflow?

davidberard98 commented on torch/csrc/jit/codegen/cuda/interface.cpp:
```diff
@@ -26,6 +26,32 @@ namespace jit {
+ public:
+  LoadingNvfuserLibrary() {
+    static c10::once_flag load_nvfuser_lib_flag;
+    c10::call_once(load_nvfuser_lib_flag, [&] {
```
nit: do we need c10::call_once if this constructor is being called statically?

davidberard98 replied:
@ngimel there's a bug in std::call_once https://github.com/pytorch/pytorch/pull/78051

malfet commented on nvfuser/CMakeLists.txt:
```diff
@@ -0,0 +1,323 @@
+if(NOT BUILD_NVFUSER)
+  return()
+else()
+  cmake_minimum_required(VERSION 3.0)
```
Have you tested it on 3.0? Perhaps we should use 3.18 as head PyTorch?

jjsjann123 replied:
Good point. I'll bump this up to 3.18 as with PyTorch then.

malfet commented on nvfuser/CMakeLists.txt:
```diff
@@ -0,0 +1,323 @@
+if(NOT BUILD_NVFUSER)
+  return()
+else()
```
Just `endif` here?
```suggestion
endif()
```

malfet commented on build_variables.bzl:
```diff
@@ -675,107 +675,6 @@ libtorch_cuda_core_sources = [
     "torch/csrc/jit/codegen/fuser/cuda/fused_kernel.cpp",
     "torch/csrc/profiler/stubs/cuda.cpp",
     "torch/csrc/autograd/functions/comm.cpp",
-    "torch/csrc/jit/codegen/cuda/arith.cpp",
```
Hmm, looks like this change makes nvfuser inaccessible to bazel (and therefore internally) Is this intentional?
cc: @davidberard98

jjsjann123 replied:
We are aware of this and do plan on adding bazel build back in future PR, have a comment in the PR

> Since we moved nvfuser into a cmake build, we effectively disabled bazel build for nvfuser. This could impact internal workload at Meta, so we need to put support back. cc'ing @vors

jjsjann123 committed:
review comments

jjsjann123 committed:
enable nvfuser tests

jjsjann123 commented:
Addressed review comments.

Added a check for loaded nvfuser library when turning on the fuser, so there will an expected error popping up for build where nvfuser is not available. https://github.com/pytorch/pytorch/pull/89621/commits/c282c5526878ab9f64edb9c34359c1a98be0ac6a

jjsjann123 commented:
Looks like CI does struggle with finding the library.

`2022-12-15T10:15:28.8955423Z /opt/conda/lib/python3.10/site-packages/torch/bin/nvfuser_tests: error while loading shared libraries: libnvfuser_codegen.so: cannot open shared object file: No such file or directory`

jjsjann123 commented:
```
2022-12-15T11:06:09.0376928Z + ln -sf /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch.so /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_cuda_linalg.so /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_global_deps.so /opt/conda/lib/python3.10/site-packages/torch/lib/libtorch_python.so /opt/conda/lib/python3.10/site-packages/torch/lib/libtorchbind_test.so /opt/conda/lib/python3.10/site-packages/torch/bin
...
2022-12-15T11:06:33.9718768Z + /opt/conda/lib/python3.10/site-packages/torch/bin/nvfuser_tests --gtest_output=xml:test/test-reports/cpp-unittest/test_libtorch/nvfuser_tests.xml
```

Looks like we just forgot to link `libnvfuser_codegen.so` there and seems to be another script issue. Will find out where it is and patch it.

jjsjann123 committed:
copying libnvfuser_codegen.so to binary folder for tests

jjsjann123 commented:
oops. I'm confused now, looks like cmake is generating `.so` library for windows as well. https://github.com/pytorch/pytorch/actions/runs/3706588201/jobs/6282702110

I'll revert the library search then.

jjsjann123 committed:
reverting win library search extension

jjsjann123 committed:
Revert "reverting win library search extension"

This reverts commit 0688597f036f9aee504d1800fb5a25ba582cb608.

jjsjann123 committed:
fixing win library names

jjsjann123 commented:
win failure looks real, so I do need to change library extension for win build, but my previous attempt forgot to remove `lib` prefix.

macos failure looks unrelated?!?! since it's cpu only build? But it does feel strange that the error wasn't here before we enable nvfuser tests.
failure: https://github.com/pytorch/pytorch/actions/runs/3707474157/jobs/6284349130
build: https://github.com/pytorch/pytorch/actions/runs/3707474157/jobs/6284012632

jjsjann123 commented:
Yay~~ apparently everything still works even with cpp tests nvfuser_test enabled~~ @davidberard98 

Also enabled error checking on enablement. So there should be CI failure on internals. We can add skip for those if we are ok with merging this PR with bazel build for nvfuser disabled. I'm playing with bazel build already, will try to figure out how to get bazel enabled...

jjsjann123 committed:
Merge remote-tracking branch 'upstream/viable/strict' into HEAD

jjsjann123 committed:
Merge remote-tracking branch 'upstream/viable/strict' into build_install

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into build_install

jjsjann123 committed:
moving nvfuser to third_party instead of top level package

jjsjann123 committed:
removing BAZEL rule for generated nvfuser header

jjsjann123 committed:
updating hipify path for new nvfuser location

jjsjann123 committed:
Merge remote-tracking branch 'upstream/viable/strict' into HEAD

jjsjann123 commented:
inductor failure doesn't look real. I'm merging master/viable/strict to try my luck there.

jjsjann123 commented:
So trunk is just broken. :cry: I'll wait until we have a clean commit and try rebase then.

jjsjann123 committed:
Merge commit '7c58f1d4e8d8bf88353eee88a30fb700e7077250' into HEAD

jjsjann123 committed:
Merge remote-tracking branch 'upstream/viable/strict' into HEAD

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jjsjann123 commented:
hmmm.. That's strange... I saw git complaining about a conflict, but merge master doesn't raise anything... I'm testing build locally now.

jjsjann123 commented:
Hahaha, clearly putting 2023/1/1 for forward_backward skip shows how optimistic I was about this PR. will bump it up to Feb 1st.

jjsjann123 committed:
bump forward_backward compatibility skip date

jjsjann123 commented:
CI looks green again now~~ except the import mismatch. :tada:

vors commented on BUILD.bazel:
```diff
@@ -1573,25 +1573,7 @@ cc_library(
-    srcs = ["torch/csrc/jit/codegen/cuda/tools/stringify_file.py"],
-)
-
-generated_nvfuser_hdrs = ["generated_" + hdr for hdr in libtorch_nvfuser_generated_headers]
```
Sorry the PR is massive, so I may something obvious -- are we dropping nvfuser from the bazel build completely? That doesn't seem right...

jjsjann123 replied:
Yes we are doing that in this PR.

> Since we moved nvfuser into a cmake build, we effectively disabled bazel build for nvfuser. This could impact internal workload at Meta, so we need to put support back. cc'ing @vors

We do plan to put it back though. Bazel build is... not easily tested and verified, so I'm trying to do that in a separate PR, otherwise, this one is going to be dragging forever... :cry:

jjsjann123 replied:
David started a separate PR disabling nvfuser in bazel workload #91836 , just to make sure that it's safe to do so in this PR.

davidberard98 replied:
@jjsjann123 @vors actually, my PR is only disabling nvfuser in the _internal_ bazel build; not in the OSS bazel build.

jjsjann123 replied:
Thx for clarifying~

jjsjann123 commented:
Looks like there's some conflicts now, should I bump this PR to upstream/master?

davidberard98 commented:
we can probably wait until next week unless you suspect there will be any new failures that will pop up

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jjsjann123 commented:
Looks like I got a bad upstream commit. I'll keep an eye on the CI hud and grab a clean commit when it gets green again.

davidberard98 commented:
recommend using the `viable/strict` branch instead of `master` next time to avoid that issue :)

jjsjann123 committed:
Merge remote-tracking branch 'upstream/viable/strict' into HEAD

jjsjann123 committed:
updating cpp standard

jjsjann123 commented:
Looks like the build failure has been patched.. Sorry that I forgot to update c++14 flag... :stuck_out_tongue:

davidberard98 commented on build_variables.bzl:
```diff
@@ -18,34 +18,34 @@ GENERATED_LAZY_TS_CPP = [
-    "torch/csrc/jit/codegen/cuda/runtime/warp.cu",
-    "torch/csrc/jit/codegen/cuda/runtime/warp_rocm.cu",
-    "torch/csrc/jit/codegen/cuda/runtime/welford.cu",
+    "nvfuser/runtime/array.cu",
```
@jjsjann123 i think these need to be updated to third_party/nvfuser ?

jjsjann123 committed:
updating nvfuser runtime file path in bzl rule

jjsjann123 commented:
windows CI seems strangely flaky.

```
ModuleNotFoundError: No module named 'typing_extensions'
```

jjsjann123 commented:
hmmm. can't seem to get the flaky CI to pass... I'll keep trying.

jjsjann123 committed:
Merge commit 'f0e3c4929b7430e37dd96cda011a0ef72d3156d3' into HEAD

jjsjann123 commented:
hmm. looks like upstream/viable/strict shows failing tests (inductor) in HUD log. I tried to grab one that at least looks green across the column. :crossed_fingers:

jjsjann123 commented:
I think I need to bump this for import/land, but HUD showing all later commits with some failures (even the commit pointed by `pytorch/viable/strict`).
I'll keep an eye out for that. In the mean time, feel free to ping me if there's any specific commits that you want me to pull @davidberard98

jjsjann123 committed:
Merge remote-tracking branch 'origin/viable/strict' into HEAD

davidberard98 reviewed (approved):
stamp so I can attempt to land internally (don't use pytorchbot to merge this)

jjsjann123 commented:
> stamp so I can attempt to land internally (don't use pytorchbot to merge this)

So excited!!! :partying_face:

jjsjann123 commented:
:partying_face:
